cache/cache_test.go

```diff
@@ -0,0 +1,203 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"testing"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+func TestValidateWatchRange(t *testing.T) {
+	type tc struct {
+		name        string
+		watchKey    string
+		opts        []clientv3.OpOption
+		cachePrefix string
+		wantErr     bool
+	}
+
+	tests := []tc{
+		{
+			name:        "single key",
+			watchKey:    "/a",
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
+			name:        "single key outside prefix returns error",
+			watchKey:    "/z",
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "explicit range",
+			watchKey:    "/a",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/b")},
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
+			name:        "exact prefix range",
+			watchKey:    "/a",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/b")},
+			cachePrefix: "/a",
+			wantErr:     false,
+		},
+		{
+			name:        "prefix subrange",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/foo/a")},
+			cachePrefix: "/foo",
+			wantErr:     false,
+		},
+		{
+			name:        "reverse range returns error",
+			watchKey:    "/b",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/a")},
+			cachePrefix: "",
+			wantErr:     true,
+		},
+		{
+			name:        "empty range returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/foo")},
+			cachePrefix: "",
+			wantErr:     true,
+		},
+		{
+			name:        "range strating below cache prefix returns error",
+			watchKey:    "/a",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/foo")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "range ending above cache prefix returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/z")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "range encompassing cache prefix returns error",
+			watchKey:    "/a",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/z")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "range crossing prefixEnd returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/z")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "range crossing prefixEnd returns error",
+			watchKey:    "/foo",
+			opts:        []clientv3.OpOption{clientv3.WithRange("/z")},
+			cachePrefix: "/foo",
+			wantErr:     true,
+		},
+		{
+			name:        "empty prefix",
+			watchKey:    "",
+			opts:        []clientv3.OpOption{clientv3.WithPrefix()},
+			cachePrefix: "",
+			wantErr:     false,
+		},
+		{
+			name:        "empty prefix with cachePrefix",
```

Comment:
This should return error, cache will return only objects with "/foo" prefix, while the watcher wants all objects.